src/librustc_typeck/outlives/explicit.rs

```diff
@@ -0,0 +1,81 @@
+// Copyright 2013 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#![allow(unused)]
+#[allow(dead_code)]
+
+use rustc::ty::{self, CratePredicatesMap, TyCtxt};
+use rustc::ty::maps::Providers;
+use std::rc::Rc;
+use util::nodemap::FxHashMap;
+use hir::map as hir_map;
+use rustc::hir;
+use rustc::hir::itemlikevisit::ItemLikeVisitor;
+use rustc::hir::def_id::{CrateNum, DefId, LOCAL_CRATE};
+
+pub fn explicit_map<'a, 'tcx: 'a>(tcx: TyCtxt<'a, 'tcx, 'tcx>, crate_num: CrateNum)
+        -> FxHashMap<DefId, Predicate> {
+    assert_eq!(crate_num, LOCAL_CRATE);
+
+    //let mut explicit_outlives_predicates = map();
+    let explicit_predicates = FxHashMap();
+
+    let mut visitor = ExplicitVisitor {
+        tcx,
+        explicit_predicates
+    };
+    //iterate over all the crates
+    tcx.hir.krate().visit_all_item_likes(&mut visitor);
+
+    explicit_predicates
+}
+
+
+pub struct ExplicitVisitor {
+    tcx: TyCtxt<'a, 'tcx, 'tcx>,
+    explicit_predicates: FxHashMap<DefId, Vec<String>>,
+}
+
+impl<'a, 'tcx, 'v> ItemLikeVisitor<'v> for ExplicitVisitor {
+    fn visit_item(&mut self, item: &hir::Item) {
+
+        // Compute a map from each struct/enum/union S to the **explicit**
+        // outlives predicates (`T: 'a`, `'a: 'b`) that the user wrote.
+        // Typically there won't be many of these, except in older code where
+        // they were mandatory. Nonetheless, we have to ensure that every such
+        // predicate is satisfied, so they form a kind of base set of requirements
+        // for the type.
+        //for def_id in all_types() {
+        //    let explicit_predicates = tcx.explicit_predicates(def_id);
+        let def_id = DefId.local(item.hir_id.owner);
```

Comment:
@nikomatsakis  I dont know how to retrieve a DefId here from Item. Can you suggest a method from this struct. https://github.com/rust-lang/rust/blob/master/src/librustc/hir/mod.rs#L1816